### PR TITLE
feature: implement simple spinlock test

### PR DESCRIPTION
**Pull request by TaiJuWu**
## Summary
feature: implement simple spinlock test

Two threads use share variable and the variable should be same as excepted.

## Impact
ostest if config spinlock
## Testing
./tools/configure.sh -l qemu-armv8a:nsh_smp
ostest

sim:ostest

### CODE REVIEW

File: testing/ostest/spinlock.c [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+ ****************************************************************************/
+
+static spinlock_t lock = SP_UNLOCKED;
+
+static pthread_t g_thread1;
+static pthread_t g_thread2;
+
+static int g_result = 0;
+
+static FAR void thread_spinlock(FAR void *parameter)
```
Comment (xiaoxiang781216) [inline testing/ostest/spinlock.c:47]: `FAR void` to `FAR void *`

File: testing/ostest/spinlock.c [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+void spinlock_test(void)
+{
+  int status;
+
+  status = pthread_create(&g_thread1, NULL, (void *)thread_spinlock, NULL);
```
Comment (xiaoxiang781216) [inline testing/ostest/spinlock.c]: remove `(void *)`

File: testing/ostest/spinlock.c [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  if (status != 0)
+    {
+      printf("spinlock_test: ERROR pthread_create failed, status=%d\n",
+              status);
+      ASSERT(false);
+    }
+
+  status = pthread_create(&g_thread2, NULL, (void *)thread_spinlock, NULL);
```
Comment (xiaoxiang781216) [inline testing/ostest/spinlock.c]: ditto

File: testing/ostest/spinlock.c [hunk 4]
```diff
@@ -46,10 +46,14 @@ static int g_result = 0;
 
 static FAR void thread_spinlock(FAR void *parameter)
 {
+  int pid = *(int *)parameter;
```
Comment (xiaoxiang781216) [inline testing/ostest/spinlock.c:49]: `int *` to `FAR int *`

File: testing/ostest/spinlock.c [hunk 5]
```diff
@@ -46,10 +46,14 @@ static int g_result = 0;
...
+
   for (int i = 0; i < 10; i++)
     {
       spin_lock(&lock);
+      printf("pid %d get lock g_result:%d\n", pid, g_result);
```
Comment (xiaoxiang781216) [inline testing/ostest/spinlock.c:54]: it wrong to call printf inside a spin lock

File: testing/ostest/spinlock.c (hunk 4 above)
Comment (TaiJuWu) [reply testing/ostest/spinlock.c]: Done.

File: testing/ostest/spinlock.c (hunk 5 above)
Comment (TaiJuWu) [reply testing/ostest/spinlock.c:57]: I got it and I add a new rw_spinlock test for future.

**Comment by xiaoxiang781216** [comment]
please fix:
```
====================================================================================
Configuration/Tool: esp32s3-devkit/ksta_softap
------------------------------------------------------------------------------------
  Cleaning...
make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
  Configuring...
make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
  Building NuttX...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0

100  340k  100  340k    0     0  1044k      0 --:--:-- --:--:-- --:--:-- 1044k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  496k  100  496k    0     0  1872k      0 --:--:-- --:--:-- --:--:-- 5169k
xtensa-esp32s3-elf-ld: /github/workspace/sources/nuttx/staging//libapps.a(spinlock.c.github.workspace.sources.apps.testing.ostest.o):(.literal.thread_native_spinlock+0x10): undefined reference to `spin_lock'
xtensa-esp32s3-elf-ld: /github/workspace/sources/nuttx/staging//libapps.a(spinlock.c.github.workspace.sources.apps.testing.ostest.o): in function `thread_native_spinlock':
spinlock.c:(.text.thread_native_spinlock+0x1b): undefined reference to `spin_lock'
make[1]: *** [Makefile:82: nuttx_user.elf] Error 1
make[1]: Target 'all' not remade because of errors.
make: *** [tools/Unix.mk:526: nuttx] Error 2
make: Target 'all' not remade because of errors.
/github/workspace/sources/nuttx/tools/testbuild.sh: line 370: /github/workspace/sources/nuttx/../nuttx/nuttx.manifest: No such file or directory
  Normalize esp32s3-devkit/ksta_softap
```

**Comment by TaiJuWu** [comment]
> please fix:
> 
> ```
> ====================================================================================
> Configuration/Tool: esp32s3-devkit/ksta_softap
> ------------------------------------------------------------------------------------
>   Cleaning...
> make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
> make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
>   Configuring...
> make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
> make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
>   Building NuttX...
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> 
> 100  340k  100  340k    0     0  1044k      0 --:--:-- --:--:-- --:--:-- 1044k
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> 
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> 100  496k  100  496k    0     0  1872k      0 --:--:-- --:--:-- --:--:-- 5169k
> xtensa-esp32s3-elf-ld: /github/workspace/sources/nuttx/staging//libapps.a(spinlock.c.github.workspace.sources.apps.testing.ostest.o):(.literal.thread_native_spinlock+0x10): undefined reference to `spin_lock'
> xtensa-esp32s3-elf-ld: /github/workspace/sources/nuttx/staging//libapps.a(spinlock.c.github.workspace.sources.apps.testing.ostest.o): in function `thread_native_spinlock':
> spinlock.c:(.text.thread_native_spinlock+0x1b): undefined reference to `spin_lock'
> make[1]: *** [Makefile:82: nuttx_user.elf] Error 1
> make[1]: Target 'all' not remade because of errors.
> make: *** [tools/Unix.mk:526: nuttx] Error 2
> make: Target 'all' not remade because of errors.
> /github/workspace/sources/nuttx/tools/testbuild.sh: line 370: /github/workspace/sources/nuttx/../nuttx/nuttx.manifest: No such file or directory
>   Normalize esp32s3-devkit/ksta_softap
> ```

It seems nuttx repo setting error. I've check the .config in nuttx and There is CONFIG_SPINLOCK on esp32s3-devkit/ksta_softap. But the setting does not compile nuttx/spinlock.o.

Because CI/CD pull code from docker, I can't modify the code from docker.

**Comment by TaiJuWu** [comment]
> please fix:
> 
> ```
> ====================================================================================
> Configuration/Tool: esp32s3-devkit/ksta_softap
> ------------------------------------------------------------------------------------
>   Cleaning...
> make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
> make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
>   Configuring...
> make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
> make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
>   Building NuttX...
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> 
> 100  340k  100  340k    0     0  1044k      0 --:--:-- --:--:-- --:--:-- 1044k
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> 
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> 100  496k  100  496k    0     0  1872k      0 --:--:-- --:--:-- --:--:-- 5169k
> xtensa-esp32s3-elf-ld: /github/workspace/sources/nuttx/staging//libapps.a(spinlock.c.github.workspace.sources.apps.testing.ostest.o):(.literal.thread_native_spinlock+0x10): undefined reference to `spin_lock'
> xtensa-esp32s3-elf-ld: /github/workspace/sources/nuttx/staging//libapps.a(spinlock.c.github.workspace.sources.apps.testing.ostest.o): in function `thread_native_spinlock':
> spinlock.c:(.text.thread_native_spinlock+0x1b): undefined reference to `spin_lock'
> make[1]: *** [Makefile:82: nuttx_user.elf] Error 1
> make[1]: Target 'all' not remade because of errors.
> make: *** [tools/Unix.mk:526: nuttx] Error 2
> make: Target 'all' not remade because of errors.
> /github/workspace/sources/nuttx/tools/testbuild.sh: line 370: /github/workspace/sources/nuttx/../nuttx/nuttx.manifest: No such file or directory
>   Normalize esp32s3-devkit/ksta_softap
> ```

The error is related about nuttx repo setting.
I just disable spinlock config, I think it just a temporary solution.

**Comment by TaiJuWu** [comment]
> has more need fix? @TaiJuWu

No. Just one PR.

**Comment by TaiJuWu** [comment]
Is there any place which can select linker obj exclude Makefile and CmakeList?

**Comment by xiaoxiang781216** [comment]
> Is there any place which can select linker obj exclude Makefile and CmakeList? After compiling, I can see spinlock.o in nuttx but it doesn't be linked. As I know I just need to set Makefile and CmakeList and then tools can help me to compiler and link them.

Does spinlock.o expose spin_lock function?

**Comment by TaiJuWu** [comment]
> > Is there any place which can select linker obj exclude Makefile and CmakeList? After compiling, I can see spinlock.o in nuttx but it doesn't be linked. As I know I just need to set Makefile and CmakeList and then tools can help me to compiler and link them.
> 
> Does spinlock.o expose spin_lock function?

Yes, it doest.
The symbol table show below
```
/home/ubuntu/nuttxSpace/nuttx/sched/spinlock.o:     file format elf64-littleriscv

SYMBOL TABLE:
0000000000000000 l    df *ABS*	0000000000000000 spinlock.c
0000000000000000 l    d  .text	0000000000000000 .text
0000000000000000 l    d  .data	0000000000000000 .data
0000000000000000 l    d  .bss	0000000000000000 .bss
0000000000000000 l    d  .text.spin_lock	0000000000000000 .text.spin_lock
0000000000000000 l    d  .text.spin_lock_wo_note	0000000000000000 .text.spin_lock_wo_note
0000000000000000 l    d  .text.spin_trylock	0000000000000000 .text.spin_trylock
0000000000000000 l    d  .text.spin_trylock_wo_note	0000000000000000 .text.spin_trylock_wo_note
0000000000000000 l    d  .text.spin_unlock	0000000000000000 .text.spin_unlock
0000000000000000 l    d  .text.spin_unlock_wo_note	0000000000000000 .text.spin_unlock_wo_note
0000000000000000 l    d  .text.spin_setbit	0000000000000000 .text.spin_setbit
0000000000000000 l    d  .text.spin_clrbit	0000000000000000 .text.spin_clrbit
0000000000000000 l    d  .debug_info	0000000000000000 .debug_info
0000000000000000 l    d  .debug_abbrev	0000000000000000 .debug_abbrev
0000000000000000 l    d  .debug_loclists	0000000000000000 .debug_loclists
0000000000000000 l    d  .debug_aranges	0000000000000000 .debug_aranges
0000000000000000 l    d  .debug_rnglists	0000000000000000 .debug_rnglists
0000000000000000 l    d  .debug_line	0000000000000000 .debug_line
0000000000000000 l    d  .debug_str	0000000000000000 .debug_str
0000000000000000 l    d  .debug_line_str	0000000000000000 .debug_line_str
0000000000000000 l    d  .comment	0000000000000000 .comment
0000000000000000 l    d  .debug_frame	0000000000000000 .debug_frame
0000000000000000 l    d  .riscv.attributes	0000000000000000 .riscv.attributes
0000000000000000 g     F .text.spin_lock	000000000000002e spin_lock
0000000000000000         *UND*	0000000000000000 up_testset
0000000000000000 g     F .text.spin_lock_wo_note	0000000000000008 spin_lock_wo_note
0000000000000000 g     F .text.spin_trylock	0000000000000024 spin_trylock
0000000000000000 g     F .text.spin_trylock_wo_note	0000000000000008 spin_trylock_wo_note
0000000000000000 g     F .text.spin_unlock	000000000000000e spin_unlock
0000000000000000 g     F .text.spin_unlock_wo_note	0000000000000008 spin_unlock_wo_note
0000000000000000 g     F .text.spin_setbit	000000000000005a spin_setbit
0000000000000000 g     F .text.spin_clrbit	0000000000000062 spin_clrbit

```

**Comment by xiaoxiang781216** [comment]
it's strange that linker can't find it.

**Comment by TaiJuWu** [comment]
> strange

Yes, so I think is there any place which should be set.

File: testing/ostest/spinlock.c [hunk 6]
```diff
@@ -61,16 +65,19 @@ static FAR void thread_spinlock(FAR void *parameter)
 void spinlock_test(void)
 {
   int status;
+  g_result = 0;
 
-  status = pthread_create(&g_thread1, NULL, (void *)thread_spinlock, NULL);
+  status = pthread_create(&g_thread1, NULL,
+                          (void *)thread_spinlock, &g_thread1);
```
Comment (xiaoxiang781216) [inline testing/ostest/spinlock.c]: remove the cast

File: testing/ostest/spinlock.c [hunk 7]
```diff
@@ -61,16 +65,19 @@ static FAR void thread_spinlock(FAR void *parameter)
...
   if (status != 0)
     {
       printf("spinlock_test: ERROR pthread_create failed, status=%d\n",
               status);
       ASSERT(false);
     }
 
-  status = pthread_create(&g_thread2, NULL, (void *)thread_spinlock, NULL);
+  status = pthread_create(&g_thread2, NULL,
+                          (void *)thread_spinlock, &g_thread2);
```
Comment (xiaoxiang781216) [inline testing/ostest/spinlock.c]: ditto

File: testing/ostest/spinlock.c [hunk 8]
```diff
@@ -42,33 +42,38 @@ static spinlock_t lock = SP_UNLOCKED;
...
 
+#ifdef CONFIG_RW_SPINLOCK
+static pthread_t g_thread3;
+static rwlock_t rw_lock = RW_SP_UNLOCKED;
+#endif
+
 static int g_result = 0;
 
-static FAR void thread_spinlock(FAR void *parameter)
+static FAR void *thread_native_spinlock(FAR FAR void *parameter)
```
Comment (xiaoxiang781216) [inline testing/ostest/spinlock.c:52]: move spinlock change to the second patch

**Comment by acassis** [comment]
@TaiJuWu do you plan to fix the conflicts and merge this PR, or can we close it?

**Comment by TaiJuWu** [comment]
> @TaiJuWu do you plan to fix the conflicts and merge this PR, or can we close it?

I am not working on this, close it.

**Pull request closed without merging**